Explore the 'serialization' issue in getAttribute('transform')

karlcow opened the pull request:
#### 811c0028641980f4af8906e8d5843e77c7124e07
<pre>
Explore the &apos;serialization&apos; issue in getAttribute(&apos;transform&apos;)
<a href="https://bugs.webkit.org/show_bug.cgi?id=267401">https://bugs.webkit.org/show_bug.cgi?id=267401</a>
<a href="https://rdar.apple.com/121178816">rdar://121178816</a>

Reviewed by NOBODY (OOPS!).

Fix the serialization of the SVG transform attribute.
After setting, for example,
    svg.createSVGTransform().setScale(-2, -4)
on an SVG element, a g.getAttribute(&apos;transform&apos;) would return.
    scale(-2 -4)
This patch fixes it so it would return:
    scale(-2, -4)
as specified.
This is normalizing and fixing some tests. But we should be adding WPT
tests for this too, because this fix doesn&apos;t break any WPT tests, which
means this is not tested in WPT.

* LayoutTests/svg/dom/SVGTransformList-basics-expected.txt:
* LayoutTests/svg/dom/SVGTransformList-basics.xhtml:
* LayoutTests/svg/dom/SVGViewSpec-expected.txt:
* LayoutTests/svg/dom/SVGViewSpec-multiple-views-expected.txt:
* LayoutTests/svg/dom/SVGViewSpec-multiple-views.html:
* LayoutTests/svg/dom/SVGViewSpec.html:
* LayoutTests/svg/dom/viewspec-parser-1-expected.txt:
* LayoutTests/svg/transforms/negative-scale-value-expected.txt:
* LayoutTests/svg/transforms/negative-scale-value.html:
* Source/WebCore/svg/SVGTransformValue.h:
(WebCore::SVGTransformValue::appendFixedPrecisionNumbers):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/811c0028641980f4af8906e8d5843e77c7124e07

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/36583 "Passed style check") | [❌ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/15522 "Hash 811c0028 for PR 23492 does not build (failure)") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/38808 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/39244 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/32803 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/37817 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/17996 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/12603 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/31416 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/37144 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/13085 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/32362 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/11453 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/11470 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/32619 "Found 1 new test failure: http/tests/websocket/tests/hybi/url-parsing.html (failure)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/40489 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/33150 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/32962 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/37398 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/11721 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/9564 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/35505 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/13405 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/32281 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/12144 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/12606 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/1aa5fa73eb446029d711a3497489e03f6529179c. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/23492

karlcow commented:
Let's start as a DRAFT PR, to see if this is breaking any current tests.
Then I should add a series of WPT tests to cover this type of issues.

karlcow commented:
So the patch is breaking some tests for cases like this one.

```
data:text/html,<svg><circle id="circle1" cx="40" cy="40" r="40" fill="green" transform="scale(2, 2) translate(10 10)"/></svg>
```

So for an input: `transform="scale(2, 2) translate(10 10)"`
The serialization for `getAttribute('transform')`:

* Safari before this patch: `scale(2 2) translate(10 10)`
* Safari after this patch: `scale(2, 2) translate(10, 10)`
* Firefox: `scale(2, 2) translate(10 10)`

Firefox leaves it as-is.

On the other hand `setScale(x, y)` and `setTranslate(x, y)` requires the comma and will always be serialized with the comma. 

That said with our new serialization we do not regress any WPT, so I would be in favor of fixing the tests in `LayoutTests/svg/`

Ahmad-S792 commented on Source/WebCore/svg/SVGTransformValue.h:
```diff
@@ -246,8 +246,10 @@ class SVGTransformValue {
     {
-        if (builder.length() && builder[builder.length() - 1] != '(')
+        if (builder.length() && builder[builder.length() - 1] != '(') {
+            builder.append(',');
```
I think we can do single 'builder.append' with `builder.append(', ')`.

karlcow replied:
it fails. I tried. It seems to require a unique character. But indeed there should be a better way.

Ahmad-S792 replied:
Fair. My next would to search 'build.append(' in searchfox.org (wubkat) to see if we have usage of it with two spaces. I might have seen it in CSS area. Hence, I suggested it.

karlcow replied:
```
In file included from /Users/karlcow/code/OpenSource/WebKitBuild/Debug/DerivedSources/WebCore/unified-sources/UnifiedSource97.cpp:1:
In file included from /Users/karlcow/code/OpenSource/Source/WebCore/accessibility/AccessibilityRenderObject.cpp:70:
In file included from /Users/karlcow/code/OpenSource/Source/WebCore/rendering/svg/legacy/LegacyRenderSVGShape.h:30:
In file included from /Users/karlcow/code/OpenSource/Source/WebCore/rendering/svg/legacy/LegacyRenderSVGModelObject.h:34:
In file included from /Users/karlcow/code/OpenSource/Source/WebCore/svg/SVGElement.h:25:
In file included from /Users/karlcow/code/OpenSource/Source/WebCore/svg/properties/SVGAnimatedPropertyImpl.h:42:
In file included from /Users/karlcow/code/OpenSource/Source/WebCore/svg/SVGTransformList.h:25:
In file included from /Users/karlcow/code/OpenSource/Source/WebCore/svg/SVGTransform.h:30:
/Users/karlcow/code/OpenSource/Source/WebCore/svg/SVGTransformValue.h:250:28: error: multi-character character constant [-Werror,-Wmultichar]
            builder.append(', ');
1 error generated.
In file included from /Users/karlcow/code/OpenSource/WebKitBuild/Debug/DerivedSources/WebCore/unified-sources/UnifiedSource96.cpp:6:
In file included from /Users/karlcow/code/OpenSource/Source/WebCore/accessibility/AccessibilityNodeObject.cpp:78:
In file included from /Users/karlcow/code/OpenSource/Source/WebCore/svg/SVGElement.h:25:
In file included from /Users/karlcow/code/OpenSource/Source/WebCore/svg/properties/SVGAnimatedPropertyImpl.h:42:
In file included from /Users/karlcow/code/OpenSource/Source/WebCore/svg/SVGTransformList.h:25:
In file included from /Users/karlcow/code/OpenSource/Source/WebCore/svg/SVGTransform.h:30:
/Users/karlcow/code/OpenSource/Source/WebCore/svg/SVGTransformValue.h:250:28: error: multi-character character constant [-Werror,-Wmultichar]
            builder.append(', ');
** BUILD FAILED **
```

karlcow replied:
Oh maybe `builder.append(", ");` instead of `builder.append(', ');`

Ahmad-S792 replied:
> Oh maybe `builder.append(", ");` instead of `builder.append(', ');`

I am already trying it on my local build.

<img width="895" alt="image" src="https://github.com/WebKit/WebKit/assets/52317531/aa5c6c8f-cff8-409e-a7a4-e612a4a8f179">

Ahmad-S792 replied:
It compiles.

Ahmad-S792 replied:
> It compiles.

@karlcow: `builder.append(", ");`

annevk commented:
Note that `getAttribute()` doesn't perform serialization (or at least not observably so) so this needs a better description as to what it is fixing.

karlcow committed:
Explore the 'serialization' issue in getAttribute('transform')
https://bugs.webkit.org/show_bug.cgi?id=267401
rdar://121178816

Reviewed by NOBODY (OOPS!).

Fix the serialization of the SVG transform attribute.
After setting, for example,
    svg.createSVGTransform().setScale(-2, -4)
on an SVG element, a g.getAttribute('transform') would return.
    scale(-2 -4)
This patch fixes it so it would return:
    scale(-2, -4)
as specified.
This is normalizing and fixing some tests. But we should be adding WPT
tests for this too, because this fix doesn't break any WPT tests, which
means this is not tested in WPT.

* LayoutTests/svg/dom/SVGTransformList-basics-expected.txt:
* LayoutTests/svg/dom/SVGTransformList-basics.xhtml:
* LayoutTests/svg/dom/SVGViewSpec-expected.txt:
* LayoutTests/svg/dom/SVGViewSpec-multiple-views-expected.txt:
* LayoutTests/svg/dom/SVGViewSpec-multiple-views.html:
* LayoutTests/svg/dom/SVGViewSpec.html:
* LayoutTests/svg/dom/viewspec-parser-1-expected.txt:
* LayoutTests/svg/transforms/negative-scale-value-expected.txt:
* LayoutTests/svg/transforms/negative-scale-value.html:
* Source/WebCore/svg/SVGTransformValue.h:
(WebCore::SVGTransformValue::appendFixedPrecisionNumbers):

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/811c0028641980f4af8906e8d5843e77c7124e07. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/23492

karlcow commented:
@shallawa 

> Your PR is targeting the third case where the transform property of the SVGGraphicsElement, which is SVGAnimatedTransformList, is changed through its baseVal member. In this case we have to sync it with the transform attribute so we have to serialize the SVGTransformList of the baseVal to a string.

**yes**

The initial issue is expressed here:

```
var svg = document.createElementNS('http://www.w3.org/2000/svg', 'svg');
var g = document.createElementNS('http://www.w3.org/2000/svg', 'g');
var transform = svg.createSVGTransform();
transform.setScale(-2, -4);
g.transform.baseVal.appendItem(transform)
g.getAttribute('transform')
```

And the results are:
```
         IN             OUT
Safari:  scale(-2, -4)  scale(-2 -4)
Firefox: scale(-2, -4)  scale(-2, -4)
Chrome:  scale(-2, -4)  scale(-2 -4)
```

And as @annevk said I probably fixed the wrong thing [getAttribute](https://dom.spec.whatwg.org/#dom-element-getattribute) should just send what is in the attribute, but somehow `setScale` serialized the value with space instead of comma. 

https://svgwg.org/svg2-draft/coords.html#TransformProperty links to 

> User agents must support the [transform](https://svgwg.org/svg2-draft/coords.html#TransformProperty) property and presentation attribute as defined in [[css-transforms-1](https://svgwg.org/svg2-draft/refs.html#ref-css-transforms-1)].

which has a section about [6. The SVG [transform](https://www.w3.org/TR/SVG11/coords.html#TransformAttribute) Attribute](https://drafts.csswg.org/css-transforms/#svg-transform)

The diagram says that it can be `comma-wsp`


> The title of your PR should be something like: Fix the serialization of the SVGTransformList property of the 
> SVGGraphicsElement when it is synced with the transform attribute by separating the numbers by commas.

Yes the PR is at a draft stage, it just picked up the title of the bug opened by @Ahmad-S792 so far. This will be changed if/when the PR is finalized.

> 1. What makes FireFox serialization more favorable than WebKit and Chrome serialization?

As you pointed out, in the third case, Safari has modified the value stored in baseVal and doesn't match the one of the document. getAttribute is not supposed to modified the value from the document.

> 2. Since all browsers support separating with spaces and commas  for the `transform` attribute, what makes separating with comma more favorable than separating with spaces?

Second `setScale(x, y)` becomes `scale(x y)` in Safari and `scale(x, y)` in Firefox. 

[setScale, setTranslate](https://svgwg.org/svg2-draft/coords.html#__svg__SVGTransform__setScale) seems to imply that it should be serialized as `scale(x, y)`

karlcow commented:
Note to myself

* [SVGTransform::setScale](https://searchfox.org/wubkat/rev/54c72ceeb7d7e8ea549af9d7c89913e7ab28208e/Source/WebCore/svg/SVGTransform.h#112-120)
* [SVGTransformValue::setScale](https://searchfox.org/wubkat/rev/54c72ceeb7d7e8ea549af9d7c89913e7ab28208e/Source/WebCore/svg/SVGTransformValue.h#153-159)
* [AffineTransform::makeScale](https://searchfox.org/wubkat/rev/54c72ceeb7d7e8ea549af9d7c89913e7ab28208e/Source/WebCore/platform/graphics/transforms/AffineTransform.h#193-196)

Also 

[SVGTransformList::parseGeneric](https://searchfox.org/wubkat/rev/54c72ceeb7d7e8ea549af9d7c89913e7ab28208e/Source/WebCore/svg/SVGTransformList.cpp#65-91) has a section about parsing the value.

annevk commented:
It doesn't seem like serialization is actually defined. "Seems to imply" is not really good enough to make a change as Chromium could reasonably object, no? This should be discussed in some standards setting first.

karlcow commented:
Diving a bit more into this:

Let's head to https://svgwg.org/svg2-draft/coords.html#__svg__SVGTransform__setScale

> The setTranslate, setScale, setRotate, setSkewX and setSkewY methods are used to set the [SVGTransform](https://svgwg.org/svg2-draft/coords.html#InterfaceSVGTransform) to a new transform function value. When one of these methods is called, the following steps are run:

Then steps 2 and let's stick to `scale` for the sake of the exercise:

> Set the [SVGTransform](https://svgwg.org/svg2-draft/coords.html#InterfaceSVGTransform) object's [value](https://svgwg.org/svg2-draft/coords.html#TransformValue) to a new transform function value, depending on which method was called:

> setScale(sx, sy)
>    the new transform function value is scale(sx, sy)

Step 3

> Set the components of the [SVGTransform](https://svgwg.org/svg2-draft/coords.html#InterfaceSVGTransform) object's [matrix object](https://svgwg.org/svg2-draft/coords.html#TransformMatrixObject) to match the new transform function [value](https://svgwg.org/svg2-draft/coords.html#TransformValue).

Step 4

> if the [SVGTransform](https://svgwg.org/svg2-draft/coords.html#InterfaceSVGTransform) object [reflects an element of a presentation attribute value](https://svgwg.org/svg2-draft/coords.html#TransformMode), then [reserialize](https://svgwg.org/svg2-draft/types.html#TermReserialize) the reflected attribute.

Ah talking about serialization…
There is also a paragraph later starting with 
> If the [DOMMatrix](https://www.w3.org/TR/geometry-1/#DOMMatrix) [reflects an SVGTransform](https://svgwg.org/svg2-draft/coords.html#MatrixMode), then:

and the step 3 talks about serialization too:

> If the [SVGTransform](https://svgwg.org/svg2-draft/coords.html#InterfaceSVGTransform) object [reflects an element of a presentation attribute value](https://svgwg.org/svg2-draft/coords.html#TransformMode), then [reserialize](https://svgwg.org/svg2-draft/types.html#TermReserialize) the reflected attribute.

Both point to the same section for reserialize, let see…
https://svgwg.org/svg2-draft/types.html#TermReserialize

They mention [SVGAnimatedTransformList](https://svgwg.org/svg2-draft/coords.html#InterfaceSVGAnimatedTransformList)
> [Reserialize](https://svgwg.org/svg2-draft/types.html#TermReserialize) the content attribute using value's baseVal member.

Then a bit lower the steps for [SVGTransformList](https://svgwg.org/svg2-draft/coords.html#InterfaceSVGTransformList)

> 1. Let elements be the list of values in value.
> > The values will be [SVGNumber](https://svgwg.org/svg2-draft/types.html#InterfaceSVGNumber), [SVGLength](https://svgwg.org/svg2-draft/types.html#InterfaceSVGLength), [DOMPoint](https://www.w3.org/TR/geometry-1/#dom-dompoint) or [SVGTransform](https://svgwg.org/svg2-draft/coords.html#InterfaceSVGTransform) objects, or DOMString values, depending on value's type.
> 2. Let serialized elements be a list of strings, where each string is formed based on the corresponding value in elements and its type:
>
> [cutting different types]
>
> a [SVGTransform](https://svgwg.org/svg2-draft/coords.html#InterfaceSVGTransform) object
> The string is the [serialization](https://drafts.csswg.org/css-transforms-1/#serialization-of-transform-functions) of the [SVGTransform](https://svgwg.org/svg2-draft/coords.html#InterfaceSVGTransform) object's [matrix object](https://svgwg.org/svg2-draft/coords.html#TransformMatrixObject).
>
> [cut]
> 3. Set the content attribute to a string consisting of the strings in **serialized elements joined and separated by single U+0020 SPACE characters**.

So the step 3 here is if there is `scale` and `translate` separate them by a space such as `transform="scale(sx, sy) translate(tx, tu))"`.

Let's go back to step 2:

> The string is the [serialization](https://drafts.csswg.org/css-transforms-1/#serialization-of-transform-functions) of the [SVGTransform](https://svgwg.org/svg2-draft/coords.html#InterfaceSVGTransform) object's [matrix object](https://svgwg.org/svg2-draft/coords.html#TransformMatrixObject). 

which leads to

> 3.1. Serialization of [<transform-function>](https://drafts.csswg.org/css-transforms-1/#typedef-transform-function)s
> 
> To serialize the [<transform-function>](https://drafts.csswg.org/css-transforms-1/#typedef-transform-function)s, serialize as per their individual grammars, in the order the grammars are written in, avoiding [<calc()>](https://drafts.csswg.org/css-values-4/#funcdef-calc) expressions where possible, avoiding <calc()> transformations, omitting components when possible without changing the meaning, **joining space-separated tokens with a single space, and following each serialized comma with a single space**.

So the spec is asking to serialize it as two values separated by a comma and a space.

@annevk, @shallawa what do you think?

annevk commented:
Thanks, yeah, I think that's sufficient to attempt to follow.

karlcow commented:
Oh I started to create testcase, and this is going to be fun…

`setScale(2, 2)`

* Safari: `scale(2 2)` or `scale(2, 2)` with my current patch
* Firefox: `scale(2)`
* Chrome: `scale(2 2)`

karlcow commented:
* according to the serialization rules above  `setScale(2, 2)` should return `scale(2, 2)`
* and `transform="scale(2)"` too because of [transform functions](https://drafts.csswg.org/css-transforms-1/#two-d-transform-functions)

> If the second parameter is not provided, it takes a value equal to the first.

karlcow commented:
Oh and this is different for `translate()`

> If `<ty>` is not provided, ty has zero as a value.

karlcow commented:
```
data:text/html,<svg><g><circle r="100" transform="scale(2)"/></svg></g>
```

then

```
const circle = document.querySelector("circle");
circle.getAttribute('transform');
"scale(2)" 
circle.transform.baseVal.getItem(0).setScale(3, 4)
circle.getAttribute('transform');
"scale(3, 4)" in Firefox, "scale(3 4)" in Safari
```

longsonr commented:
The usual rule in serialisation is to serialise to something as short as possible. That's why Firefox serialises transform="scale(2, 2)" to scale(2)

There are special rules for rotate (to omit the x and y if they are not needed) and translate.

https://searchfox.org/mozilla-central/source/dom/svg/SVGTransform.cpp#25

karlcow commented:
> The usual rule in serialisation is to serialise to something as short as possible. That's why Firefox serialises transform="scale(2, 2)" to scale(2)
> 
> There are special rules for rotate (to omit the x and y if they are not needed) and translate.
> 
> https://searchfox.org/mozilla-central/source/dom/svg/SVGTransform.cpp#25

@longsonr is there a section of one of the specs which says this?

longsonr commented:
There certainly used to be something like this in the CSS specification: https://www.w3.org/TR/2011/WD-cssom-20110712/#serializing-css-values

> Where CSS component values of the value can be omitted without changing the meaning of the value (e.g. initial values in shorthand properties), omit them. If this would remove all the values, then include the first allowed value.

https://github.com/w3c/csswg-drafts/issues/878 talks about reinstating those principles somehow.

https://www.w3.org/Graphics/SVG/Test/20061213/htmlEmbedHarness/full-styling-css-02-b.html demanded a certain minimal serialization which prompted Firefox and probably other browsers to implement it. Release versions of Chrome and Safari currently seem to pass this test too so we'd probably not want to break that.

As far as the Firefox implementation goes we have https://bugzilla.mozilla.org/show_bug.cgi?id=736791 and we've certainly had these serialisation code rules since before Firefox 9 It's hard to search before that as the code used to be in CVS and history was not preserved with the switch to mercurial.

karlcow commented:
The draft of CSSOM for [serializing CSS values](https://drafts.csswg.org/cssom/#serializing-css-values)

> If component values can be omitted or replaced with a shorter representation without changing the meaning of the value, omit/replace them.


But what is happening currently is:

```
data:text/html,<svg><g><circle r="100" transform="scale(2)"/></svg></g>
```

and Safari, Chrome:

```js
const circle = document.querySelector("circle");
circle.getAttribute('transform');
# return  "scale(2)"
circle.transform.baseVal.getItem(0).setScale(3, 3)
circle.getAttribute('transform');
# return "scale(3 3)"
```

Firefox:
```js
const circle = document.querySelector("circle");
circle.getAttribute('transform');
# return  "scale(2)"
circle.transform.baseVal.getItem(0).setScale(3, 3)
circle.getAttribute('transform');
# return "scale(3)"
```

with my patch for WebKit (currently but could be changed to match Firefox.)
```js
const circle = document.querySelector("circle");
circle.getAttribute('transform');
# return  "scale(2)"
circle.transform.baseVal.getItem(0).setScale(3, 3)
circle.getAttribute('transform');
# return "scale(3, 3)"
```

longsonr commented:
So in markup you'd be able to distinguish transform="scale(2)" from transform="scale(2, 2)" wouldn't you?

https://www.w3.org/TR/SVG11/coords.html#TransformAttribute said

> [scale](https://www.w3.org/TR/SVG11/coords.html#ScalingDefined)(<sx> [<sy>]), which specifies a scale operation by sx and sy. If <sy> is not provided, it is assumed to be equal to <sx>.

Suggesting that you take scale(2), convert it to scale(2, 2) and then process that.

This text was moved to the [css-transforms specification](https://www.w3.org/TR/css-transforms-1/#funcdef-transform-scale) in SVG 2 and that says...

> If the second parameter is not provided, it takes a value equal to the first.

Suggesting the same i.e. scale(2) and scale(2, 2) in markup should be indistinguishable. I think Firefox's implementation matches my reading of the specification's intent more closely so I'd be very reluctant to change it.

longsonr commented:
> And just as a reminder for stating the obvious:
> 

Ah but that's the attribute and not the property. You don't normalise that, you just output the text you were given. You could do this for instance...

> data:text/html,<svg><g><circle r="100" transform="I'm a teapot"/></svg></g>
> ```js
> const circle = document.querySelector("circle");
> circle.getAttribute('transform');
> // return  "I'm a teapot"
> ```

Safari does return "I'm a teapot" in that case doesn't it?

karlcow commented:
In Firefox, @longsonr 

* `setScale(2, 2)` returns `scale(2)`
* **but** `setTranslate(2, 2)` returns `translate(2)`

is it expected?

longsonr commented:
> In Firefox, @longsonr
> 
>     * `setScale(2, 2)` returns `scale(2)`
> 
>     * **but** `setTranslate(2, 2)` returns `translate(2)`
> 
> 
> is it expected?

No, that's correct for scale but not for translate and the serialisation logic does not seem to do that for translate. https://searchfox.org/mozilla-central/source/dom/svg/SVGTransform.cpp#25

Please provide a complete testcase and/or raise a bug in bugzilla.

karlcow commented:
> No, that's correct for scale but not for translate and the serialisation logic does not seem to do that for translate. https://searchfox.org/mozilla-central/source/dom/svg/SVGTransform.cpp#25
> Please provide a complete testcase and/or raise a bug in bugzilla.

I messed up my cut and paste above. just fixed it :) So Firefox is correct. Thanks for the pointer.

karlcow commented:
The rules are defined here for future references. 
https://drafts.csswg.org/css-transforms-1/#two-d-transform-functions
